### ISSUE TITLE
Error with logging while running run_all_fl.sh. TypeError: cannot unpack non-iterable NoneType object

### ISSUE THREAD

**Issue by Linardos** [Project-MONAI/tutorials#764]
I went through the instructions of https://github.com/Project-MONAI/tutorials/tree/main/federated_learning/breast_density_challenge README and ran ./run_all_fl.sh as instructed. The process seems to have completed normally; however, when I tried visualizing on the tensorboard I got nothing. I checked the folder and it seems result_server/run_1/tb_events is empty. Is there something I need to adjust for the tb_events to be stored? I am running the latest version of the code.

Attached find the log.txt generated after running the bash script

[log.txt](https://github.com/Project-MONAI/tutorials/files/9010319/log.txt)

**Comment by Linardos** [comment]
Apparently, the execution crashes somehow at the point where it attempts to store global_acc, global_kappa. Here's a segment from site-1_log.txt:

![image](https://user-images.githubusercontent.com/26694607/177193784-9ee89f8a-0f39-42ab-9207-b4f1b4c8967c.png)

I'm also attaching that txt for your reference. Not sure what could be wrong. I followed the instructions to the T and repeated the process from scratch 3 times but always get the same issue. The error is apparently the same in all txt files. 
I changed the name of this issue as it's clearly more general than tb_events storage.

[site-1_log.txt](https://github.com/Project-MONAI/tutorials/files/9041111/site-1_log.txt)

**Comment by Nic-Ma** [comment]
Hi @YuanTingHsieh ,

Do you know any NVFLare expert who can help take a look at this issue as Holger is PTO now?

Thanks in advance.

**Comment by holgerroth** [comment]
Sorry @Linardos for the late reply. The error stems from here in [site-1_log.txt](https://github.com/Project-MONAI/tutorials/files/9041111/site-1_log.txt)
```
2022-06-29 09:22:12,650 - MammoLearner - CRITICAL - [run=1]: /data/dataset_blinded_site-1.json does not exist!
2022-06-29 09:22:12,650 - LearnerExecutor - ERROR - [run=1]: learner initialize exception: [Errno 2] No such file or directory: '/data/dataset_blinded_site-1.json'
```

**Comment by holgerroth** [comment]
Please follow the steps here https://github.com/Project-MONAI/tutorials/tree/main/federated_learning/breast_density_challenge#11-download-example-data

**Comment by Linardos** [comment]
Huh, weird. I suppose in one of the runs something got mixed up and I sent you a wrong log. However, the overall issue is not there. When I follow the steps correctly (I verified the data/dataset_blinded_site-1 in fact exists) the issue TypeError: cannot unpack non-iterable NoneType object persists. Here's the corresponding log (I couldn't find a CRITICAL error here):
[site-1_log (1).txt](https://github.com/Project-MONAI/tutorials/files/9047803/site-1_log.1.txt)

The only kind of ERRORs I identified look like this:
`2022-07-04 17:38:22,439 - LearnerExecutor - ERROR - [run=1, peer=fl_project, peer_run=1, task_name=train, task_id=433ce423-6942-432b-b2d3-9c1de9a232d7]: learner execute exception: BEFORE_TRAIN_VALIDATE failed: cannot unpack non-iterable NoneType object
`

**Comment by holgerroth** [comment]
You have to look for the first error in the log. There's still seems to be a problem with the I/O:
```
Loading dataset:   0%|          | 0/559 [00:00<?, ?it/s]2022-07-04 17:36:17,225 - LearnerExecutor - ERROR - [run=1]: learner initialize exception: applying transform <monai.transforms.io.dictionary.LoadImaged object at 0x7f349e433580>
```
After that, the validation logic cannot compute the metric scores which results in the `cannot unpack non-iterable NoneType object` error.

Seems the error is already happening when the CacheDataset are being created in the `initialize()` function of the `MammoLearner`. 

Can you try debugging the mammo_learner.py inside the container by following [these instructions](https://github.com/Project-MONAI/tutorials/tree/main/federated_learning/breast_density_challenge#21-debugging-the-learning-algorithm)?

**Comment by Linardos** [comment]
I see. I tried it out and it seems nvflare had not been properly installed. Apparently it's failing to install something called tenseal. I raised an issue in the nvflare library instead:
https://github.com/NVIDIA/NVFlare/issues/706

**Comment by holgerroth** [comment]
The tenseal error doesn't seem to be related to the above error messages. Are you using docker to run this example?

**Comment by Linardos** [comment]
I was referring to the debugger. I used python3 code/pt/learners/mammo_learner.py and noticed the nvflare issues. That's resolved now anyway and everything is imported correctly; however, when I run it now this error is presented:
```
(bdenv) akis-linardos@gandalfthegray:~/BDChallenge/tutorials/federated_learning/breast_density_challenge$ python code/pt/learners/mammo_learner.py
Testing MammoLearner...
Traceback (most recent call last):
  File "code/pt/learners/mammo_learner.py", line 662, in <module>
    engine = MockClientEngine()
  File "code/pt/learners/mammo_learner.py", line 630, in __init__
    self.fl_ctx_mgr = FLContextManager(
TypeError: __init__() got an unexpected keyword argument 'run_num'
```



When I used the containerized version `./run_docker_debug.sh` it starts running and then it doesn't find a file it's looking for. However, that file doesn't seem to be included in the files downloaded by python3 ./code/pt/utils/download_datalists_and_predictions.py
![image](https://user-images.githubusercontent.com/26694607/177422213-595f57de-b84d-45ee-81f5-299fb1fcc5d8.png)
Instead only phase 2 files are downloaded when I use that script:
![image](https://user-images.githubusercontent.com/26694607/177422404-7d842be7-f456-4cb2-8b08-2afed9b12a74.png)

**Comment by holgerroth** [comment]
Notice that this reference implementation was written for [nvflare 2.0.16](https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/federated_learning/breast_density_challenge/Dockerfile#L20). I think you are using a newer version.

**Comment by Linardos** [comment]
I downgraded to 2.0.16
Now both the docker and the script get the same error of not finding the phase1 file.
![image](https://user-images.githubusercontent.com/26694607/177547489-b81b4031-ca17-4069-97ac-85e1056bbf4a.png)
I run the downloading script again. It doesn't download any phase1 file:
`
(bdenv) akis-linardos@gandalfthegray:~/BDChallenge/tutorials/federated_learning/breast_density_challenge$ python ./code/pt/utils/download_datalists_and_predictions.py
`

`
(bdenv) akis-linardos@gandalfthegray:~/BDChallenge/tutorials/federated_learning/breast_density_challenge/data$ ls
dataset_blinded_phase2_site-1.json  dataset_blinded_phase2_site-3.json  dataset_blinded_site-2.json  dataset_site-1.json  dataset_site-3.json
dataset_blinded_phase2_site-2.json  dataset_blinded_site-1.json         dataset_blinded_site-3.json  dataset_site-2.json
`

**Comment by holgerroth** [comment]
Thank you for reporting this inconsitency.  There's not supposed to be a phase1 dataset. I prepared a fix here https://github.com/Project-MONAI/tutorials/pull/772/files

The actual dataset used during FL is specified in `config_fed_client.json` [here](https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/federated_learning/breast_density_challenge/code/configs/mammo_fedavg/config/config_fed_client.json#L5).

**Comment by Linardos** [comment]
Cool! However the file needed one more fix to overcome this error: You have the value "inside_container=True" (line 647) even when running the script outside the container:
![image](https://user-images.githubusercontent.com/26694607/177644292-1561f252-6164-4e94-9f9e-8c9c5c00497d.png)
So it's looking for data in the root instead of in the current directory. I set it to False.

However, more errors are popping up now when I try to run the debugger:

```
Traceback (most recent call last):
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/transforms/transform.py", line 89, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/transforms/transform.py", line 53, in _apply_transform
    return transform(parameters)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/transforms/io/dictionary.py", line 121, in __call__
    data = self._loader(d[key], reader)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/transforms/io/array.py", line 194, in __call__
    img = reader.read(filename)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/image_reader.py", line 574, in read
    img = np.load(name, allow_pickle=True, **kwargs_)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/numpy/lib/npyio.py", line 407, in load
    fid = stack.enter_context(open(os_fspath(file), "rb"))
FileNotFoundError: [Errno 2] No such file or directory: 'data/preprocessed/Mass-Training_P_00004_LEFT_MLO.npy'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "code/pt/learners/mammo_learner.py", line 667, in <module>
    learner.initialize(parts={}, fl_ctx=fl_ctx)
  File "code/pt/learners/mammo_learner.py", line 243, in initialize
    self.train_dataset = CacheDataset(
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/dataset.py", line 722, in __init__
    self.set_data(data)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/dataset.py", line 747, in set_data
    self._cache = _compute_cache()
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/dataset.py", line 736, in _compute_cache
    return self._fill_cache()
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/dataset.py", line 763, in _fill_cache
    return list(p.imap(self._load_cache_item, range(self.cache_num)))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 865, in next
    raise value
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/data/dataset.py", line 776, in _load_cache_item
    item = apply_transform(_xform, item)
  File "/home/akis-linardos/bdenv/lib/python3.8/site-packages/monai/transforms/transform.py", line 113, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.io.dictionary.LoadImaged object at 0x7f34d4a8e340>

```

**Comment by holgerroth** [comment]
For debugging purposes, you can either inside or outside the container. That flag is just for utility so that the path should be automatically updated. When setting it too false, it assume you use

**Comment by Linardos** [comment]
I didn't remove the ./ I only set inside_container to False:
![image](https://user-images.githubusercontent.com/26694607/177659840-1008dfed-fda4-4118-aa1c-5bbe28324359.png)

The problem is not in the path. The folder "preprocessed" does not exist. The script `python ./code/pt/utils/download_datalists_and_predictions.py` does not download it:
![image](https://user-images.githubusercontent.com/26694607/177660208-51840727-e4b6-4b55-ac6f-d57dcc5a7cac.png)

Here's the download script. 
![image](https://user-images.githubusercontent.com/26694607/177660520-a47d31af-0973-4829-bfcf-eb8a17e63d91.png)

Something is missing. I visited these sites manually and didn't find a "preprocessed" folder.
